pkg/gadgettracermanager/stream/stream.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stream
+
+import (
+	"sync"
+	"time"
+)
+
+const (
+	MAX_BUFFERED_LINES = 100
+)
+
+type TimestampedLine struct {
+	Line      string
+	Timestamp time.Time
+}
+
+// GadgetStream
+type GadgetStream struct {
+	mu sync.RWMutex
+
+	previousLines []TimestampedLine
+
+	// subs contains a list of subscribers
+	subs []chan TimestampedLine
+
+	closed bool
+}
+
+func NewGadgetStream() *GadgetStream {
+	return &GadgetStream{
+		subs: make([]chan TimestampedLine, 0),
+	}
+}
+
+func (g *GadgetStream) Subscribe() chan TimestampedLine {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.closed {
+		return nil
+	}
+
+	ch := make(chan TimestampedLine, MAX_BUFFERED_LINES)
+	for _, l := range g.previousLines {
+		ch <- l
+	}
+	g.subs = append(g.subs, ch)
+
+	return ch
+}
+
+func (g *GadgetStream) Publish(line string) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.closed {
+		return
+	}
+
+	newLine := TimestampedLine{
+		Line:      line,
+		Timestamp: time.Now(),
+	}
+
+	g.previousLines = append(g.previousLines, newLine)
+	if len(g.previousLines) > MAX_BUFFERED_LINES {
+		g.previousLines = g.previousLines[len(g.previousLines)-MAX_BUFFERED_LINES:]
+	}
```

Comment:
:thinking: how efficient is it? Is it worth to use a circular buffer instead?

Reply:
You're right. I think the memory usage would always increase because the memory at the beginning of the array is not released when redefining a slice in that way...